### PR TITLE
Stage Python tutorial

### PR DESCRIPTION
**Pull request by rachfop**
**Preview:** https://temporal-learning-git-sagas-python.preview.thundergun.io/tutorials/python/sagas/

### CODE REVIEW

File: docs/tutorials/python/sagas/index.md [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+keywords: [Python, temporal, sdk, tutorial, saga, saga pattern, rollback, compensating actions]
+tags: [Python, SDK]
+last_update:
+  date: 2024-03-06
+title: Implement the Saga Pattern in Python Using Temporal
+description: Implement the Saga Pattern in Python Using Temporal.
+image: /img/temporal-logo-twitter-card.png
+---
+
+In today's interconnected digital landscape, ensuring consistency and reliability in multi-step processes is paramount.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:13]:
```suggestion
In today's interconnected digital landscape, ensuring consistency and reliability in multi-step processes is key.
```

"paramount" might be kind of an advanced word for ESL people

File: docs/tutorials/python/sagas/index.md [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+  "result": "Voyage cancelled",
+  "user_id": "jane-smith-609592"
+}
+```
+
+In this case, the booking process was cancelled due to the invalid hotel booking ID.
+The Saga pattern ensures that any completed bookings are rolled back, maintaining a consistent state.
+
+This demonstrates how the Saga pattern with Temporal handles both successful and failing scenarios in the booking process.
+Let's summarize what you've accomplished and discuss potential next steps.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:545]: Where's the summary?

File: docs/tutorials/python/sagas/index.md [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+last_update:
+  date: 2024-03-06
+title: Implement the Saga Pattern in Python Using Temporal
+description: Implement the Saga Pattern in Python Using Temporal.
+image: /img/temporal-logo-twitter-card.png
+---
+
+In today's interconnected digital landscape, ensuring consistency and reliability in multi-step processes is paramount.
+When dealing with distributed systems, a failure in one service can lead to a domino effect, compromising the entire transaction.
+The Saga pattern offers a solution to this problem by allowing distributed transactions to be broken into smaller, manageable transactions, each with its own compensation logic in case of failure.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:15]: Maybe do a brief sentence of what compensation logic is or link to somewhere else?

File: docs/tutorials/python/sagas/index.md [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+
+These classes and constants will be used in both Activities, Workflows, and Workers.
+
+Workflows order and run the execution of Activities.
+With your Activities and shared data classes defined, the next step is to create the Workflow.
+This Workflow coordinates the execution of Activities and handle compensations to maintain consistency in case of failures.
+
+## Create the Workflow Definition
+
+In the context of Temporal Workflows, compensation refers to the actions taken to roll back a transaction if an error occurs.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:186]: I would move this sentence up to the part where you first mention compensations.

File: docs/tutorials/python/sagas/index.md [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+
+
+TASK_QUEUE_NAME = "saga-task-queue"
+```
+
+These classes and constants will be used in both Activities, Workflows, and Workers.
+
+Workflows order and run the execution of Activities.
+With your Activities and shared data classes defined, the next step is to create the Workflow.
+This Workflow coordinates the execution of Activities and handle compensations to maintain consistency in case of failures.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:182]:
```suggestion
This Workflow coordinates the execution of Activities and handles compensations to maintain consistency in case of failures.
```

File: docs/tutorials/python/sagas/index.md (hunk 2 above)
Comment (angelazhou32) [reply docs/tutorials/python/sagas/index.md:545]: Oh is the summary in the conclusion? If so, I think you're missing the next steps!

File: docs/tutorials/python/sagas/index.md [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+```
+
+For this example, if the number of attempts is less than the allowed number of attempts or if the booking ID is invalid, the Activity will raise exceptions to simulate failures, then run the corresponding undo action.
+
+By setting up these compensations, you'll ensure that your system can handle failures gracefully.
+Next, you'll focus on defining shared data classes and constants to support your Activities and Workflows.
+
+## Define Shared Data Classes and Constants
+
+Shared data classes and constants are used to pass data between Activities and Workflows.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:155]: Not just across Activities and Workflows though (e.g., client, workers). Maybe: "We will now create some shared data classes and constants that we will use through the application.'

File: docs/tutorials/python/sagas/index.md [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+    book_car_id: str
+    book_hotel_id: str
+    book_flight_id: str
+    attempts: int
+
+
+TASK_QUEUE_NAME = "saga-task-queue"
+```
+
+These classes and constants will be used in both Activities, Workflows, and Workers.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:178]:
```suggestion
These classes and constants will be used throughout Activities, Workflows, and Workers.
```
since 'both' is just for two items

File: docs/tutorials/python/sagas/index.md [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+
+Workflows order and run the execution of Activities.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:180]: I don't think you need this line 'Workflows order and run the execution of Activities.' since you say in the line after next 'this workflow coordinates the execution of Activities'..can be kind of redundant?

File: docs/tutorials/python/sagas/index.md [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+    undo_book_flight,
+    undo_book_hotel,
+)
+from saga_workflows import BookWorkflow
+from shared import TASK_QUEUE_NAME
+```
+
+
+The Worker is a crucial component that executes the defined Workflows and Activities.
+Setting up the Worker correctly will ensure that your system can process tasks efficiently and reliably.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:307]: You already have these two lines on lines 279-280

File: docs/tutorials/python/sagas/index.md [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+
+You've successfully initiated and completed your booking process using the Saga pattern with Temporal in Python.
+Next, you'll learn how to simulate errors to test the robustness of your implementation.
+
+### Simulate an error
+
+To ensure your implementation can handle failures gracefully, you will simulate a booking failure.
+This step will demonstrate how the Saga pattern with Temporal manages to rollback in case of errors.
+
+To simulate a booking failure, you can use the following `curl` command. T
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:512]: Is the 'This' written as one word or separated into two? Just checking!

File: docs/tutorials/python/sagas/index.md [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+```
+
+<!--
+The `asyncio` library is used for asynchronous operations.
+The `activity` module from the `temporalio` library provides decorators and functions for defining Activities.
+The `BookVacationInput` data class will be used to pass input data to the Activities.
+-->
+
+Next, define the `book_car`, `book_hotel`, and `book_flight` Activities.
+These Activities interact with external services.
```
Comment (angelazhou32) [inline docs/tutorials/python/sagas/index.md:62]: Should we say that we assume they are in this case? Because in the code they aren't

**Comment by angelazhou32** [comment]
Great stuff!! :D

File: docs/tutorials/python/sagas/index.md (hunk 4 above)
Comment (napcs) [reply docs/tutorials/python/sagas/index.md:186]: I partially agree although I think it's good to have it here too. I wouldn't move it so much as add a section to the introduction explaining how Temporal applications implement Sagas.

**Comment by axfelix** [comment]
Thoughts on numbered steps?

File: docs/tutorials/python/sagas/index.md (hunk 4 above)
Comment (angelazhou32) [reply docs/tutorials/python/sagas/index.md:186]: Yeah but maybe reword/reemphasize this and move this to the first section when compensations are mentioned, because I think people would benefit from an explanation of what it means :)

File: docs/tutorials/python/sagas/index.md (hunk 6 above)
Comment (napcs) [reply docs/tutorials/python/sagas/index.md:155]: I'd like to see the title of this section and the intro paragraph make a stronger case for *why* you are doing this. This feels like something I must do for Temporal, not something that's a good practice.

Think about how you'd explain this to someone at a booth demo. Why are we doing this?

File: docs/tutorials/python/sagas/index.md [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+TASK_QUEUE_NAME = "saga-task-queue"
+```
+
+These classes and constants will be used in both Activities, Workflows, and Workers.
+
+Workflows order and run the execution of Activities.
+With your Activities and shared data classes defined, the next step is to create the Workflow.
+This Workflow coordinates the execution of Activities and handle compensations to maintain consistency in case of failures.
+
+## Create the Workflow Definition
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:184]: Again here, this title feels too "temporal"y. You're defining the business process using a temporal workflow. Try to talk in general developer terms as much as you can.

File: docs/tutorials/python/sagas/index.md [hunk 13]
```diff
@@ -0,0 +1,551 @@
...
+
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:185]: There's something missing here that explains what the workflow is and why we're building it. You jump into the compensation stuff right away.

File: docs/tutorials/python/sagas/index.md [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+```python
+from datetime import timedelta
+from temporalio import workflow
+from temporalio.common import RetryPolicy
+
+with workflow.unsafe.imports_passed_through():
+    from activities import BookVacationInput, book_car, book_flight, book_hotel
+```
+
+Next, create the `BookWorkflow` class and define the compensation actions, as well as the functions that execute your core logic: `book_car`, `book_hotel`, and `book_flight`.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:205]: You do an excellent job of explaining how you track the compensating activities after the code. But I think it's worth explaining the general pattern here. You're going to call each activity and keep track of the compensations you need to run if they fail so that you can execute the appropriate ones at the end. I don't think you need more than that because your description after the code that explains how you do it is excellent.

File: docs/tutorials/python/sagas/index.md [hunk 15]
```diff
@@ -0,0 +1,551 @@
...
+Each Activity Execution includes a retry policy to handle transient errors.
+If any Activity fails, the `except` block catches the exception and executes the compensation activities in reverse order to undo the previously completed steps.
+This ensures the system is returned to a consistent state. The retry policy specifies how to handle retries for each Activity, including non-retryable error types and retry intervals.
+
+Having defined the Workflow, you're now ready to set up the Worker that will execute these Workflows and Activities.
+
+## Define the Worker
+
+The Worker is a crucial component that executes the defined Workflows and Activities.
+Setting up the Worker correctly will ensure that your system can process tasks efficiently and reliably.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:280]: be mindful here - what is "correctly"? I might omit this sentence.

File: docs/tutorials/python/sagas/index.md [hunk 16]
```diff
@@ -0,0 +1,551 @@
...
+```
+
+The `Client.connect()` line connects to the Temporal server running on `localhost` at port `7233`.
+This can be modified to run a Worker on Temporal Cloud.
+
+The `Worker` is initialized with the client, the task queue name, the list of Workflows, and the list of Activities.
+This setup ensures the Worker knows which tasks to listen for and execute.
+The `await worker.run()` line starts the Worker, making it ready to receive tasks and execute the corresponding Activities and Workflows.
+
+**Start the Worker**
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:358]:
```suggestion
```

Delete. Not needed as your sentence after covers this.

File: docs/tutorials/python/sagas/index.md [hunk 17]
```diff
@@ -0,0 +1,551 @@
...
+
+To start the Worker, run the following command in your terminal:
+
+```bash
+python run_worker.py
+```
+
+Once the Worker is running, it will be ready to execute Workflows and Activities as tasks are submitted to the specified task queue.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:366]:
```suggestion
Once the Worker is running, it will be ready to execute Workflows and Activities as tasks are submitted to the specified Task Queue.
```

File: docs/tutorials/python/sagas/index.md [hunk 18]
```diff
@@ -0,0 +1,551 @@
...
+
+Now that the Worker is set up and running, you can set up the Client to initiate the booking process.
+
+## Create the Client
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:370]: This is more than a client. This is 
```suggestion
## Create the Flask API to book trips
```

File: docs/tutorials/python/sagas/index.md [hunk 19]
```diff
@@ -0,0 +1,551 @@
...
+
+The Client is responsible for initiating the Workflow and handling the booking process.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:372]: This intro is talking about Temporal things - talk to developers in general. They've got all this booking logic they built with Temporal Workflows and Activities, and it's able to do the compensations, but it's not useful until you integrate it into some kind of application.

Really drive home the connection between "their world" and "our world."

File: docs/tutorials/python/sagas/index.md [hunk 20]
```diff
@@ -0,0 +1,551 @@
...
+Setting up the Client will allow you to interact with the Temporal server and trigger the booking Workflow.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:373]: Give a general overview of the API you're creating. If it's just a single endpoint that accepts POSt requests, that's fine, but it's worth saying that up front. Give them the roadmap of what they're building before you jump to the code.

File: docs/tutorials/python/sagas/index.md [hunk 21]
```diff
@@ -0,0 +1,551 @@
...
+
+Create a new file named `run_workflow.py`:
+
+First, import the necessary modules, including `uuid`, Flask, and Temporal `Client`.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:377]: why do I need uuid

File: docs/tutorials/python/sagas/index.md [hunk 22]
```diff
@@ -0,0 +1,551 @@
...
+
+from flask import Flask, jsonify, request
+from temporalio.client import Client
+
+from activities import BookVacationInput
+from saga_workflows import BookWorkflow
+from shared import TASK_QUEUE_NAME
+```
+
+Initialize the Flask app and set up the dependency injection for the Temporal client.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:390]: This could use better explanation. Why do I care about dependency injection, and why do I need to do it? 

What thing are we *really* doing here? We're making the Temporal Client available to all of our Flask routes in a way that doesn't create a new instance of the Temporal client on each web request. This lets you share resources.

The dependency injection is just an implementation detail. Keep asking "why am I doing this" and then add the answer to the tutorial. :)

File: docs/tutorials/python/sagas/index.md [hunk 23]
```diff
@@ -0,0 +1,551 @@
...
+
+```python
+app = Flask(__name__)
+
+
+async def get_temporal_client():
+    return await Client.connect("localhost:7233")
+```
+
+Define a route to handle the booking process.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:400]: I'd recommend explaining the data that this request will receive.

File: docs/tutorials/python/sagas/index.md [hunk 24]
```diff
@@ -0,0 +1,551 @@
...
+This route will accept a `POST` request, extract the necessary data from the request, initiate the Workflow, and return the result.
+
+```python
+@app.route("/book", methods=["POST"])
+async def book_vacation():
+    # Extract data from the request
+    user_id = f'{request.json.get("name").replace(" ", "-").lower()}-{str(uuid.uuid4().int)[:6]}'
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:407]: I would extract this to a separate function. Then you can give it a "generate_unique_username" name and explain the logic in a separate code block above this workflow definition. There's a lot going on here, and this logic may change.

File: docs/tutorials/python/sagas/index.md [hunk 25]
```diff
@@ -0,0 +1,551 @@
...
+
+A `BookVacationInput` object is created with the extracted data, which will be passed to the Workflow.
+
+The Temporal client is obtained using the `get_temporal_client()` function.
+
+The Workflow is executed using `client.execute_workflow()`, passing the input object and other required parameters.
+Based on the result of the Workflow execution, a response is prepared and returned.
+If the booking process is cancelled, the response indicates this. Otherwise, it provides details about the booked car, hotel, and flight.
+
+**Start the Client**
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:461]:
```suggestion
```

Omit.

File: docs/tutorials/python/sagas/index.md [hunk 26]
```diff
@@ -0,0 +1,551 @@
...
+
+Now to start the Client, run the following command in your terminal:
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:463]: Don't I already have my terminal blocked with a worker?

We need to tell people to open a new window, switch to the project dir, and then do the start process.

File: docs/tutorials/python/sagas/index.md [hunk 27]
```diff
@@ -0,0 +1,551 @@
...
+    "attempts": 5,
+    "car": "valid-car-id",
+    "hotel": "valid-hotel-id",
+    "flight": "valid-flight-id"
+}'
+```
+
+**Output**
+
+The expected output should be a JSON response similar to the following:
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:490]:
```suggestion
You'll see a JSON response similar to the following:
```

File: docs/tutorials/python/sagas/index.md [hunk 28]
```diff
@@ -0,0 +1,551 @@
...
+-d '{
+    "name": "John Doe",
+    "attempts": 5,
+    "car": "valid-car-id",
+    "hotel": "valid-hotel-id",
+    "flight": "valid-flight-id"
+}'
+```
+
+**Output**
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:488]:
```suggestion
```

omit

File: docs/tutorials/python/sagas/index.md [hunk 29]
```diff
@@ -0,0 +1,551 @@
...
+  "result": "Booked car: valid-car-id Booked hotel: valid-hotel-id Booked flight: valid-flight-id",
+  "user_id": "john-doe-184942"
+}
+```
+
+
+You've successfully initiated and completed your booking process using the Saga pattern with Temporal in Python.
+Next, you'll learn how to simulate errors to test the robustness of your implementation.
+
+### Simulate an error
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:507]:
```suggestion
## Simulate an error
```

File: docs/tutorials/python/sagas/index.md [hunk 30]
```diff
@@ -0,0 +1,551 @@
...
+-d '{
+    "name": "Jane Smith",
+    "attempts": 3,
+    "car": "valid-car-id",
+    "hotel": "invalid-hotel-id",
+    "flight": "valid-flight-id"
+}'
+```
+
+The value `invalid` will trigger an exception, causing the booking to rollback.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:527]:
```suggestion
The value `invalid` will trigger an exception, causing the booking to roll back.
```

File: docs/tutorials/python/sagas/index.md [hunk 31]
```diff
@@ -0,0 +1,551 @@
...
+
+**Output**
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:529]:
```suggestion
```

omit

File: docs/tutorials/python/sagas/index.md [hunk 32]
```diff
@@ -0,0 +1,551 @@
...
+
+The expected output should be a JSON response similar to the following:
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:531]:
```suggestion
The output in your terminal will be a JSON response similar to the following:
```

File: docs/tutorials/python/sagas/index.md [hunk 33]
```diff
@@ -0,0 +1,551 @@
...
+
+```json
+{
+  "cancelled": true,
+  "result": "Voyage cancelled",
+  "user_id": "jane-smith-609592"
+}
+```
+
+In this case, the booking process was cancelled due to the invalid hotel booking ID.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:541]: How can we verify that these compensations ran? This looks like a place to open up the temporal web ui and show the event history to show they worked.

File: docs/tutorials/python/sagas/index.md (hunk 2 above)
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:545]:
```suggestion
```

Omit. No need for a transition prior to the conclusion.

File: docs/tutorials/python/sagas/index.md [hunk 34]
```diff
@@ -0,0 +1,551 @@
...
+The Saga pattern ensures that any completed bookings are rolled back, maintaining a consistent state.
+
+This demonstrates how the Saga pattern with Temporal handles both successful and failing scenarios in the booking process.
+Let's summarize what you've accomplished and discuss potential next steps.
+
+## Conclusion
+
+In this guide, you have implemented the Saga pattern using Temporal in Python to handle distributed transactions for booking services.
+By following this guide, you now have a robust system that can gracefully handle failures and ensure data consistency across multiple services.
+This implementation can be extended to various other use cases where multi-step processes need reliable and scalable orchestration.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:551]: Passive voice. 
```suggestion
You can extend this implementation to other use cases where multi-step processes need reliable and scalable orchestration.
```

File: docs/tutorials/python/sagas/index.md [hunk 35]
```diff
@@ -0,0 +1,551 @@
...
+In this case, the booking process was cancelled due to the invalid hotel booking ID.
+The Saga pattern ensures that any completed bookings are rolled back, maintaining a consistent state.
+
+This demonstrates how the Saga pattern with Temporal handles both successful and failing scenarios in the booking process.
+Let's summarize what you've accomplished and discuss potential next steps.
+
+## Conclusion
+
+In this guide, you have implemented the Saga pattern using Temporal in Python to handle distributed transactions for booking services.
+By following this guide, you now have a robust system that can gracefully handle failures and ensure data consistency across multiple services.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:550]:
```suggestion
By following this guide, you now have a framework for your applications that can gracefully handle failures and ensure data consistency across multiple services.
```

We don't have a robust system here. We haven't handled some edge cases. We've given folks a starting point though.

**Review by napcs — CHANGES_REQUESTED** [review]

File: docs/tutorials/python/sagas/index.md [hunk 36]
```diff
@@ -0,0 +1,586 @@
...
+title: Build a trip booking application in Python
+description: Implement the Saga Pattern in Python using Temporal.
+image: /img/temporal-logo-twitter-card.png
+---
+
+When dealing with distributed systems, a failure in one service can lead to a domino effect, compromising the entire transaction.
+The Saga pattern offers a solution to this problem by allowing distributed transactions to be broken into smaller, manageable transactions, each with its own compensation logic in case of failure.
+
+The Saga pattern is a design pattern that provides a mechanism to manage long-running transactions and ensure data consistency across multiple services.
+Instead of a single monolithic transaction, the Saga pattern breaks the transaction into smaller, manageable steps (Activities), each step is executed sequentially, and if a step fails, previous steps are undone with a compensating step.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:17]: It would be good to break this sentence down a little. Perhaps after (Activities)?

File: docs/tutorials/python/sagas/index.md [hunk 37]
```diff
@@ -0,0 +1,586 @@
...
+
+Temporal orchestrates long-running transactions, automatically compensating for failures.
+The compensation, combined with the guarantee that the method will complete execution, makes this method a reliable, long-running transaction.
+
+In this guide, you'll build a Flask API that uses Temporal to manage the booking process for cars, hotels, and flights.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:22]: nit: With this guide instead of In this guide.

File: docs/tutorials/python/sagas/index.md [hunk 38]
```diff
@@ -0,0 +1,586 @@
...
+image: /img/temporal-logo-twitter-card.png
+---
+
+When dealing with distributed systems, a failure in one service can lead to a domino effect, compromising the entire transaction.
+The Saga pattern offers a solution to this problem by allowing distributed transactions to be broken into smaller, manageable transactions, each with its own compensation logic in case of failure.
+
+The Saga pattern is a design pattern that provides a mechanism to manage long-running transactions and ensure data consistency across multiple services.
+Instead of a single monolithic transaction, the Saga pattern breaks the transaction into smaller, manageable steps (Activities), each step is executed sequentially, and if a step fails, previous steps are undone with a compensating step.
+
+Temporal orchestrates long-running transactions, automatically compensating for failures.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:19]: Two things here: First, I'm not sure if we're still supposed to use "orchestration" (cc @MasonEgger ) and second, it would be a bonus to expand this to explain in a little more depth. There's a lot happening in this sentence.

File: docs/tutorials/python/sagas/index.md [hunk 39]
```diff
@@ -0,0 +1,586 @@
...
+    return f"Undoing booking of hotel: {input.book_hotel_id}"
+
+
+@activity.defn
+async def undo_book_flight(input: BookVacationInput) -> str:
+    print(f"Undoing booking of flight: {input.book_flight_id}")
+    return f"Undoing booking of flight: {input.book_flight_id}"
+```
+
+For this example, if the number of attempts is less than the allowed number of attempts or if the booking ID is invalid, the Activity will raise exceptions to simulate failures, then run the corresponding undo action.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:146]: This would benefit from breaking the text down into shorter sentences, but I'm not sure exactly how you'd do it other than maybe The activity checks if the number.... If so, it raises...??

File: docs/tutorials/python/sagas/index.md [hunk 40]
```diff
@@ -0,0 +1,586 @@
...
+The `try` block attempts to execute each booking Activity (`book_car`, `book_hotel`, `book_flight`) in sequence.
+Each Activity Execution includes a retry policy to handle transient errors.
+If any Activity fails, the `except` block catches the exception and executes the compensation activities in reverse order to undo the previously completed steps.
+This ensures the system returns to a consistent state. The retry policy specifies how to handle retries for each Activity, including non-retryable error types and retry intervals.
+
+Having defined the Workflow, you're now ready to set up the Worker that will execute these Workflows and Activities.
+
+## Define the Worker
+
+To make your booking logic, built with Temporal Workflows and Activities, functional and integrated into your application, you need to set up a Worker.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:276]: Nit: You could make the Worker more active in this sentence (or sentences). Setting up a Worker lets you integrate... A Worker is built with...

File: docs/tutorials/python/sagas/index.md [hunk 41]
```diff
@@ -0,0 +1,586 @@
...
+from activities import BookVacationInput
+from saga_workflows import BookWorkflow
+from shared import TASK_QUEUE_NAME
+```
+
+The `uuid` module is used to generate a unique ID for each booking.
+The `Flask` module is used to set up the Flask API.
+The `Client` module is used to connect to the Temporal Service.
+
+Initialize the Flask app and set up the dependency injection for the Temporal Client.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:388]: Nit: This seems to be missing a "by...." at the end of the sentence.

File: docs/tutorials/python/sagas/index.md [hunk 42]
```diff
@@ -0,0 +1,586 @@
...
+
+```python
+app = Flask(__name__)
+
+
+async def get_temporal_client():
+    return await Client.connect("localhost:7233")
+```
+
+Dependency injection for the Temporal Client is used here to ensure it is initialized once and reused, avoiding the resource-intensive process of repeatedly starting it for each booking request.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:398]: A good place to break down a long sentence is after reused.

File: docs/tutorials/python/sagas/index.md (hunk 37 above)
Comment (fairlydurable) [reply docs/tutorials/python/sagas/index.md:22]: Also, maybe a note admonition about what Flask is?

File: docs/tutorials/python/sagas/index.md [hunk 43]
```diff
@@ -0,0 +1,586 @@
...
+In this case, the booking process was cancelled due to the invalid hotel booking ID.
+The Saga pattern ensures that any completed bookings are rolled back, maintaining a consistent state.
+
+To verify the actions that took place during the booking process, you can use the Temporal Web UI.
+
+Open your instance of the Temporal Web UI in your browser and navigate to the **Workflows** tab.
+
+:::note View the Web UI
+
+If you are using the Temporal CLI, you can visit [localhost:8233/](http://localhost:8233/).
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:571]: Just a note that @MasonEgger told me that the learn site universally uses 8080 instead of 8233

File: docs/tutorials/python/sagas/index.md [hunk 44]
```diff
@@ -0,0 +1,586 @@
...
+Select the last completed Workflow, and in the **Timeline** view, you can see the Workflow Execution started, an Activity `book_car` completed, but the `book_hotel` Activity failed, retired, and then executed the `undo_book_hotel` and `undo_book_car` Activities.
+
+![View of the timeline actions in the Temporal Web UI](rollback-compenstation.png)
+
+This demonstrates how the Saga pattern with Temporal handles both successful and failing scenarios in the booking process.
+
+## Conclusion
+
+In this guide, you have implemented the Saga pattern using Temporal in Python to handle distributed transactions for booking services.
+By following this guide, you now have a framework for your applications that can gracefully handle failures and ensure data consistency across multiple services.
```
Comment (fairlydurable) [inline docs/tutorials/python/sagas/index.md:585]: nit: After following

File: docs/tutorials/python/sagas/index.md [hunk 45]
```diff
@@ -0,0 +1,586 @@
...
+    await asyncio.sleep(3)
+    if activity.info().attempt < input.attempts:
+        activity.heartbeat(
+            f"Invoking activity, attempt number {activity.info().attempt}"
+        )
+        await asyncio.sleep(3)
+        raise RuntimeError("Car service is down")
+
+    if "invalid" in input.book_car_id:
+        raise Exception("Invalid car booking, rolling back!")
```
Comment (cretz) [inline docs/tutorials/python/sagas/index.md:76]: Should not raise `Exception` directly, but something more like `RuntimeError` or something

File: docs/tutorials/python/sagas/index.md [hunk 46]
```diff
@@ -0,0 +1,586 @@
...
+    print(f"Undoing booking of flight: {input.book_flight_id}")
+    return f"Undoing booking of flight: {input.book_flight_id}"
+```
+
+For this example, if the number of attempts is less than the allowed number of attempts or if the booking ID is invalid, the Activity will raise exceptions to simulate failures, then run the corresponding undo action.
+
+By setting up these compensations, you'll ensure that your system can handle failures gracefully.
+Next, you'll focus on defining shared data classes and constants to support your Activities and Workflows.
+
+## Define Shared Data Classes and Constants
```
Comment (cretz) [inline docs/tutorials/python/sagas/index.md:151]: Seems a bit backwards to have this section after the code that depends on it (no prob of course, but putting this first may be easier than asking people to import something that won't exist until later sections)

File: docs/tutorials/python/sagas/index.md [hunk 47]
```diff
@@ -0,0 +1,586 @@
...
+                    non_retryable_error_types=["Exception"],
+                ),
+            )
+
+            # If all bookings are successful, return the output
+            return output
+        except Exception:
+            # If an error occurs, execute compensations in reverse order
+            for compensation in reversed(compensations):
+                await workflow.execute_activity(
```
Comment (cretz) [inline docs/tutorials/python/sagas/index.md:255]: You should explain what happens when the undo activity fails to confirm that behavior is intentional

File: docs/tutorials/python/sagas/index.md [hunk 48]
```diff
@@ -0,0 +1,586 @@
...
+
+from activities import (
+    book_car,
+    book_flight,
+    book_hotel,
+    undo_book_car,
+    undo_book_flight,
+    undo_book_hotel,
+)
+from saga_workflows import BookWorkflow
```
Comment (cretz) [inline docs/tutorials/python/sagas/index.md:298]: A bit inconsistent calling one of these `activities` and another `saga_workflows`. Would expect `saga_` prefix for both or neither

File: docs/tutorials/python/sagas/index.md (hunk 42 above)
Comment (cretz) [inline docs/tutorials/python/sagas/index.md:398]: > Dependency injection for the Temporal Client is used here

Where is it used here? The code in this README calls `get_temporal_client()` directly from the HTTP  handler

File: docs/tutorials/python/sagas/index.md [hunk 49]
```diff
@@ -0,0 +1,693 @@
...
+Before you begin, ensure you have the following:
+
+- Complete the [Hello World](/getting_started/python/hello_world_in_python/index.md) tutorial
+- Familiarity with asynchronous programming in Python
+- Basic understanding of microservices and distributed systems
+- The [Flask framework with async extras](https://flask.palletsprojects.com/en/2.3.x/async-await/) installed as a dependency of your project, which you can do with `pip install flask[async]`.
+
+## Create the booking functions
+
+You will start by creating the Activities for the booking process.
```
Comment (napcs) [inline docs/tutorials/python/sagas/index.md:38]: I think there's a little bit missing here. If you assume no knowledge of temporal and you land here, you might want to have something here that explains that you're going to use temporal workflows and activities to build this logic. 


```suggestion
You will start by creating the individual components for the booking process. You'll create a Temporal Workflow and a set of Activities. Activities let you interact with services, and Workflows orchestrate the Activities.
```

And I'd expect both workflows and activities to be linked to appropriate encyclopedia pages in docs. I did this in the next.js tutorial. We aren't asking for much, but this one sentence can help explain how the functions map to the core features.
